yuzu: init at 482

IvarWithoutBones opened the pull request:
###### Motivation for this change
This adds the yuzu-mainline package, a Nintendo Switch emulator.

~~The `DENABLE_VULKAN=No` CMake flag is used due to [this issue](https://github.com/yuzu-emu/yuzu/issues/3375).~~ EDIT: The issue is fixed, this is no longer required.
~~Libressl is included as its required for [this issue](https://github.com/yuzu-emu/yuzu/issues/2957).~~ EDIT: The issue is fixed, this is no longer required.

Closes #58176

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

IvarWithoutBones committed:
yuzu: init at 482

Emantor reviewed (approved):
Builds and starts on NixOS unstable.

JoshuaFern commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,27 @@
+
+  meta = with stdenv.lib; {
+    homepage = "https://yuzu-emu.org";
+    description = "An experimental open-source emulator for the Nintendo Switch";
```
I know you're using the official wording from the readme, but this description is pretty confusing. This is a Nintendo Switch emulator, not an emulator for the Switch.

How about this:
 ` description = "An experimental Nintendo Switch emulator";`

JoshuaFern commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,27 @@
+  src = fetchgit {
+    url = "https://github.com/yuzu-emu/yuzu-mainline";
+    rev = "20c42dc9ea8bdc208d9d274350af2ff347876c76";
+    sha256 = "0s6vnvbyshbrl1xxiwdrxc2jl75jyjlf4kidl88c7pwj3m2gcwss";
```
I'm getting an error with this sha256:
```
hash mismatch in fixed-output derivation '/nix/store/g3m9wqmhsp7ffvkldqkgc7fl57f92xxg-yuzu-mainline-20c42dc':
  wanted: sha256:0s6vnvbyshbrl1xxiwdrxc2jl75jyjlf4kidl88c7pwj3m2gcwss
  got:    sha256:0486nhpvl824ajksn496g0rfi9hismm8xzwdka5lplxrdmk961q2
```

JoshuaFern replied:
I think the options used are causing git to be non-deterministic, hence the changing hashes.

Specifically `leaveDotGit` , this seems to be related: https://github.com/NixOS/nixpkgs/issues/8567

JoshuaFern replied:
Oh, duh. The previous attempt at packaging this application had the same problem: https://github.com/NixOS/nixpkgs/pull/58312

JoshuaFern replied:
I've crafted my own version of yuzu-mainline in my NUR repo: https://github.com/JoshuaFern/nur-package-lab/blob/master/pkgs/misc/emulators/yuzu-mainline/default.nix

I believe this version avoids the hashing issue.

IvarWithoutBones replied:
At the moment Animal Crossing support is [a tad weird](https://www.patreon.com/posts/36005316), that's not the fault of this. I've tested Super Mario Odyssey and various [example homebrew applications](https://github.com/switchbrew/switch-examples) and they all seem to work well. 

As for the hash issue, I'll do that. I had no idea it was broken, as it worked perfectly fine locally.

As for Vulkan i haven't tried getting it to work yet, but i think it needs some specific driver packages depending on the GPU vendor. Is that something we can include in this package?

JoshuaFern replied:
> At the moment Animal Crossing support is a tad weird, that's not the fault of this. 

I noticed the Yuzu team has some kind of exclusive emulator build for donators, do you know if all the features in these builds originate from github or do they have an entirely separate and private development branch?

> As for the hash issue, I'll do that. Your solution seems to be working perfectly. I had no idea it was broken, as it worked perfectly fine locally.

You wouldn't have any idea, because every time you rebuild using the same package and hash you're just getting the data from your local nix store.

The way I test it is to change a single character of the hash and try to build, it will redownload the data and give you an error with a fresh hash.

> As for Vulkan i haven't tried getting it to work yet, but i think it needs some specific driver packages depending on the GPU vendor. Is that something we can include in this package?

I'm not sure yet, more investigation is needed.

IvarWithoutBones replied:
> I noticed the Yuzu team has some kind of exclusive emulator build for donators, do you know if all the features in these builds originate from github or do they have an entirely separate and private development branch?

This is their "Early Access" thing, its unfortunately not at all available for Linux. Patreons (their donators) get a token which they can put into the Windows-only installer, which allows them to download these builds from yuzu's private API. To comply with the GPL, this includes the source code which compiles fine on Linux about 60% of the time. Without a Windows VM its essentially not possible to use these builds.

> The way I test it is to change a single character of the hash and try to build, it will redownload the data and give you an error with a fresh hash.

Good to know, I'll keep that in mind.

JoshuaFern commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,27 @@
+  meta = with stdenv.lib; {
+    homepage = "https://yuzu-emu.org";
+    description = "An experimental open-source emulator for the Nintendo Switch";
+    license = licenses.gpl2;
```
At the tail end of the license file there's some additional licenses listed for various assets included with yuzu. Unfortunately some of these assets are under CC BY-ND 3.0, which is a non-free license.

`license = with licenses; [ gpl2 cc-by-nd-30 cc0  ];`

IvarWithoutBones commented:
Adressed review comments. I've noticed you're using `mkDerivation` instead of `wrapQtAppsHook`, is that prefered? I just did this based off of [this wiki page.](https://nixos.wiki/wiki/Qt)

JoshuaFern commented:
> Adressed review comments. I've noticed you're using `mkDerivation` instead of `wrapQtAppsHook`, is that prefered? I just did this based off of [this wiki page.](https://nixos.wiki/wiki/Qt)

I based my work off the Citra emulator package, which I'm familiar with. Both emulators originated from the same development team so I figured there's probably a lot of similarities between the two and it would be a good place to start.

On the [nixpkgs manual](https://nixos.org/nixpkgs/manual/) it states: 

> Locating runtime dependencies.  Qt applications need to be wrapped to find runtime dependencies. If you cannot use mkDerivation or mkDerivationWith above, include wrapQtAppsHook in nativeBuildInputs

So it sounds like you can use either, but `mkDerivation` is preferred.

IvarWithoutBones commented:
> So it sounds like you can use either, but `mkDerivation` is preferred.

Ah, i didn't see that wiki page. Switched over to it.

> I based my work off the Citra emulator package, which I'm familiar with. Both emulators originated from the same development team so I figured there's probably a lot of similarities between the two and it would be a good place to start.

Indeed, though from what I've heard from over at their Discord server they mostly share frontends at this point.

JoshuaFern commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,33 @@
+    done
+  '';
+
+  enableParallelBuilding = true;
```
I believe you don't need to specify `enableParallelBuilding` in this instance, from the nixpkgs manual:

> Unless set to false, some build systems with good support for parallel building including cmake, meson, and qmake will set it to true. 

So you don't need to specify this flag at all in your packages, unless to disable it. It's already on by default when it ought to be, and can be disabled when it causes problems.

JoshuaFern commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,32 @@
+    homepage = "https://yuzu-emu.org";
+    description = "An experimental Nintendo Switch emulator";
+    license = with licenses; [ gpl2 cc-by-nd-30 cc0 ];
+    maintainers = with maintainers; [ ivar ];
```
I'd like to be a co-maintainer for this package if you don't mind, `joshuafern`

IvarWithoutBones replied:
Of course :+1:

aanderse commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,32 @@
+  '';
+
+  nativeBuildInputs = [ cmake libressl ];
+  cmakeFlags = [ "-DCMAKE_BUILD_TYPE=Release" ];
```
I believe this is already implied.

IvarWithoutBones replied:
Was not aware, will remove.

aanderse commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,32 @@
+    done
+  '';
+
+  nativeBuildInputs = [ cmake libressl ];
```
I'm guessing that you want `libressl` in `buildInputs` instead.

IvarWithoutBones replied:
That's what i originally thought as well, however the issue described [here](https://github.com/yuzu-emu/yuzu/issues/2957) still occured when i did that. When moving it to `nativeBuildInputs` it works perfectly fine.

aanderse replied:
Oh, OK. So do the `cmake` scripts use `libressl` or something?

IvarWithoutBones replied:
I think so, if its not already there it pulls it as a submodule, but for some reason that doesn't fix the issue so a system provided one is needed.

jtojnar replied:
This is very suspicious.

jtojnar replied:
It seems like it is built in either case.

JoshuaFern commented:
Still working on figuring out vulkan support, slowed by compile times.

Here's the error in the logs:
`Render.Vulkan <Error> video_core/renderer_vulkan/renderer_vulkan.cpp:CreateInstance:101: Vulkan library not available`

Just tried adding `vulkan-headers` to `nativeBuildInputs` and `vulkan-loader` to `buildInputs`, that didn't work.

IvarWithoutBones commented:
I tried that last week as well, I tried the LunarG `vulkan-loader` and `vulkan-tools` as well, but that did not seem to work.

JoshuaFern commented:
Good news, I have working vulkan. Let me clean this up a bit and I'll send it your way.

JoshuaFern commented:
This seems to be working well: https://github.com/JoshuaFern/nur-package-lab/blob/master/pkgs/misc/emulators/yuzu-mainline/default.nix

JoshuaFern commented:
Do you have sound by the way? I'm playing Pokemon Mystery Dungeon DX and I'm soundless over here. Not sure if it's a problem with this game, or if the sound isn't working in general.

IvarWithoutBones commented:
> This seems to be working well: https://github.com/JoshuaFern/nur-package-lab/blob/master/pkgs/misc/emulators/yuzu-mainline/default.nix

Thanks a lot! I couldn't figure this one out myself.

> Do you have sound by the way? I'm playing Pokemon Mystery Dungeon DX and I'm soundless over here. Not sure if it's a problem with this game, or if the sound isn't working in general.

Whoops, did not notice that one as i muted it beforehand since it'd sound all crackely because of the framerate. I'll add `alsaLib`, thanks again

JoshuaFern commented:
> Whoops, did not notice that one as i muted it beforehand since it'd sound all crackely because of the framerate. I'll add `alsaLib`, thanks again

I had a similar problem with Citra, but it has another audio output engine `sdl2` which worked. Looks like we can fix that issue in that package too :)

IvarWithoutBones commented:
Disabled Vulkan with CMake if `useVulkan` is set to false, forgot to do that before.

JoshuaFern commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,41 @@
+  nativeBuildInputs = [ cmake libressl ];
+  buildInputs = [ SDL2 qtbase python2 alsaLib ]
+  ++ stdenv.lib.optionals useVulkan [ vulkan-loader vulkan-headers ];
+  cmakeFlags = stdenv.lib.optionals (!useVulkan) [ "DENABLE_VULKAN=No" ];
```
Forgot a hyphen:
  `cmakeFlags = stdenv.lib.optionals (!useVulkan) [ "-DENABLE_VULKAN=No" ];`

IvarWithoutBones replied:
I thought it adds that automatically? I wasn't using one before when i had to use this because of an issue on yuzu's end either, that worked perfectly fine.

JoshuaFern replied:
I tested building with vulkan disabled and it errored, that's how I caught the problem.

Once corrected it built correctly and functions correctly, the renderer toggle on the bottom left of the window is stuck on OpenGL and can't be toggled which proves it's working.

aanderse commented:
I hope @jtojnar doesn't mind when I ping them for reviews too often... they're just too good! :blush:

jtojnar reviewed (changes requested):
The libressl issue seems to have been fixed by https://github.com/yuzu-emu/yuzu/pull/3675. That also allows us to use more dependencies from nixpkgs so we should include that.

jtojnar commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,41 @@
+  pname = "yuzu-mainline";
+  version = "unstable-2020-04-13";
+
+  # Submodules
```
Why not use `fetchFromGithub`?

IvarWithoutBones replied:
> Why not use fetchFromGithub with fetchSubmodules?

I didn't know that made a difference, as this already fetches the submodules. Is `fetchFromGithub` preferred?

> The libressl issue seems to have been fixed by yuzu-emu/yuzu#3675. That also allows us to use more dependencies from nixpkgs so we should include that.

I wanted to check that out as well, but Linux compilation on master is currently so I couldn't really play with it. the rev this is using doesn't have the PR merged yet.

jtojnar replied:
> I didn't know that made a difference, as this already fetches the submodules. Is `fetchFromGithub` preferred?

It results in slightly cleaner expression (but uses `fetchgit` internally)-


> I wanted to check that out as well, but Linux compilation on master is currently so I couldn't really play with it. the rev this is using doesn't have the PR merged yet.

Is there an upstream issue for that?

IvarWithoutBones replied:
> It results in slightly cleaner expression (but uses fetchgit internally)-

Alright, will switch over tomorrow (No pun intended).

> Is there an upstream issue for that?

[Yes](Is there an upstream issue for that?), should be a quick and easy fix.

JoshuaFern replied:
I originally used `fetchgit` in this way because it was used in the `citra` package, and it seemed to work without the hashing issue.

You're right though, fetchFromGithub is slightly cleaner and I prefer it.

IvarWithoutBones commented:
After looking into using more libraries from nixpkgs, i found we can use `lz4`, `zstd`, `libopus`, `openssl`, and `libzip`. (Note that `fmt` won't work until #84771 is merged, yuzu requires version 6.2.0 or higher.)

I'll wait till yuzu-emu/yuzu/issues/3754 is fixed, update the git rev, include those libraries, and switch to `fetchFromGitHub`. (Not sure if i should wait for #84771 to be merged?)

JoshuaFern commented:
> I'll wait till...

Let's keep this great momentum going and get this merged, nothing is a buzzkill like waiting for upstream changes :P

IvarWithoutBones commented:
> I think what you're suggesting is sort of a hybrid approach, where we use both submodules and supply system libraries.

Yes, looking at yuzu's CMake i just picked out the packages they accept from `pkgconfig`. I've already tested this locally with a hack to make their latest master compile, works perfectly fine (though it does clone the unneeded submodules).

I'll switch to `fetchFromGitHub` now, and i think it should be ready to be merged after that if we don't want to use libraries from nixpkgs.

JoshuaFern commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,42 @@
+
+  src = fetchFromGitHub {
+    owner = "yuzu-emu";
+    repo = "${pname}";
```
`repo = pname;`

You only need to use `${pname}` if adding a variable to a string.

jtojnar replied:
I would use something like

```suggestion
    repo = "yuzu-mainline"; # they use a separate repo for mainline “branch”
```

Not really a fan of re-using `pname` in `repo` since it does not often match the project name.

JoshuaFern replied:
I think `repo = pname;` makes sense here, since if you were reusing this package to build yuzu-nightly or master you wouldn't have to replace this string manually. Also makes writing package overrides a tad simpler.

jtojnar commented:
Merged the fmt update. Do you have an expression for master I could test?

jtojnar commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,42 @@
+  };
+
+  nativeBuildInputs = [ cmake libressl ];
+  buildInputs = [ SDL2 qtbase python2 alsaLib ]
```
Does python3 not work?

JoshuaFern replied:
```
ERROR: Cannot use '/nix/store/lm0w36273v76nnph1518gwbv1g45sl8w-python3-3.7.7/bin/python', Python 2.4 or later is required.
       Note that Python 3 or later is not yet supported.
       Use --python=/path/to/python to specify a supported Python.
```

jtojnar commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,42 @@
+    sha256 = "0s9p5di9kcr8sh6iipv21b74c4dhn37an4khy4y7b9s9y8p1haim";
+  };
+
+  nativeBuildInputs = [ cmake libressl ];
```
Unless libressl is required at build time, it should go to build inputs.

JoshuaFern replied:
This has been discussed before, this is a special case.

JoshuaFern commented:
> Merged the fmt update. Do you have an expression for master I could test?

There is no explicit yuzu-master expression, it's rarely in a buildable state from my experience. That's why yuzu-mainline is preferred.

jtojnar commented:
If it is tracking a different branch than master, there should be a comment, since unstable generally refers to the default branch in the repo.

jtojnar commented on pkgs/misc/emulators/yuzu-mainline/default.nix:
```diff
@@ -0,0 +1,42 @@
+  meta = with stdenv.lib; {
+    homepage = "https://yuzu-emu.org";
+    description = "An experimental Nintendo Switch emulator";
+    license = with licenses; [ gpl2 cc-by-nd-30 cc0 ];
```
According to the readme:

```suggestion
    license = with licenses; [
      gpl2Plus
      # icons
      cc-by-nd-30 cc0
    ];
```

IvarWithoutBones commented:
> If it is tracking a different branch than master, there should be a comment, since unstable generally refers to the default branch in the repo.

The build system is a tad weird. You have yuzu-emu/yuzu, which is their main repository but it's not really usable. They assert on essentially everything, which yuzu-emu/yuzu-mainline ignores, as well as other changes which arent deemed to be fitting in their main repository but are useful to the end user. 

What i did to test system libraries locally is manually cloning yuzu-emu/yuzu-mainline, and
removing `-Werror=conversion` from src/video_core/CMakeLists.txt`. After that i just added `pkgconfig` and said libraries as both `nativeBuildInputs` and `buildInputs` respectively. That seems to be working well.

jtojnar commented:
Does setting `NIX_CFLAGS_COMPILE = "-Wno-error=conversion";` help?

IvarWithoutBones commented:
> For simplicity's sake, could we rename the package to just yuzu? I feel like yuzu-mainline is needlessly verbose.

Sounds good, done.

> Does setting NIX_CFLAGS_COMPILE = "-Wno-error=conversion"; help?

Looks like it does, though a fix is already PR'd as well: yuzu-emu/yuzu/pull/3786
I'll include it for now, we can remove it when updating the package later.

IvarWithoutBones commented:
Updated the package and switched to using system provided libraries when possible. Tested Mario Odyssey and some homebrew games with this updated version, seems to be working flawlessly.

It still clones the system provided libraries as externals, but I'm not sure of a good way around that.

IvarWithoutBones commented:
```
Could NOT find Perl (missing: PERL_EXECUTABLE)
Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
```
Looking around the codebase, it does not look like they use Doxygen for documentation (possibly a leftover from Citra? They originally forked from that). They do have a doxyfile, which also contains the only reference to perl. I think it's used as an dependency for just that.

`AudioUnit/AudioUnit.h` and `audioclient.h` seem to be Windows only, we don't need those. 
`SLES/OpenSLES.h` and `android/log.h` seem to be Android only, we don't need those either.

When using the version of `boost` in nixpkgs instead of their external, the program fails to compile. I think this is because they're using [a fork](https://github.com/yuzu-emu/ext-boost).

It won't even detect `unicorn-emu` from nixpkgs, cmake complains about not being able to find it wether we include it or not. (Possibly also because they use [a fork](https://github.com/yuzu-emu/unicorn)? Not sure). 

I've added `libpulseaudio`, `libjack2`, and `sndio`.

jtojnar commented:
Looks like started using `static_vector_options` from Boost 1.71 without bumping the minimum required version in their CMake files :frowning_woman: Nixpkgs defaults to Boost 1.69 so you need to use `boost17x` or `boost171`.

jtojnar commented:
This is a bug in FMT. Could you try the following patch: https://github.com/fmtlib/fmt/pull/1657

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/could-weve-implemented-multi-output-packages-better/6597/6

IvarWithoutBones commented:
Managed to get yuzu to detect nixpkgs provided `unicorn-emu` by passing `-DYUZU_USE_BUNDLED_UNICORN=OFF` as a CMake flag, however that [fails to compile](https://gist.github.com/IvarWithoutBones/d1d5af8c5096d7b7507fbf8a209fda6e). I'm guessing because the latest stable release (the one in nixpkgs) is `1.0.1`, from 3 years ago while yuzu needs newer functions.

I've added a few CMake flags [which their CI uses](https://github.com/yuzu-emu/yuzu/blob/3a13552db854036db4267ed21ce1e3f7f000568a/.ci/scripts/linux/docker.sh), as we would be missing out on Discord presence and webengine support otherwise. I noticed `-DYUZU_ENABLE_COMPATIBILITY_REPORTING` in there, but i thought we'd better not add that as they don't want outdated compatibility reports on fixed git revs. Note that `rapidjson` is included as it's required for Discord presence.

I also noticed we don't wrap `yuzu-cmd` (cli binary this produces as well) for Vulkan support, so I've done that.

JoshuaFern commented:
Great investigation work, I just looked into getting unicorn-emu updated. It's currently version 1.0.1, and 1.0.2 is still a release candidate. If you didn't know what Unicorn is,
> Unicorn is a lightweight, multi-platform, multi-architecture CPU emulator framework based on QEMU.

I imagine this is a pretty critical component of yuzu. The git submodule points to[ this fork](https://github.com/yuzu-emu/unicorn/tree/73f45735354396766a4bfb26d0b96b06e5cf31b2), which hasn't seen activity in years.

On a personal note, I recently ordered the parts for a PC upgrade from my ancient i7-3770 to a new shiny Ryzen 3600. I'm hoping for much faster compile times and decent speed while running and testing games on yuzu. I also preordered a Pine Phone and I'm interested in helping out with NixOS Mobile. :)

IvarWithoutBones commented:
> I imagine this is a pretty critical component of yuzu. 

Not so much, actually. They're working on getting it out as its not the fastest. Its currently just used as a fallback when [Dynarmic](https://github.com/MerryMage/dynarmic) doesn't have an instruction they need which is rarely the case. I'm sure you've seen that taking ages to compile.

> On a personal note, I recently ordered the parts for a PC upgrade from my ancient i7-3770 to a new shiny Ryzen 3600. I'm hoping for much faster compile times and decent speed while running and testing games on yuzu. I also preordered a Pine Phone and I'm interested in helping out with NixOS Mobile. :)

Cool! Sounds like it'll be a lot of fun.

jtojnar commented:
The fmt pull request was merged so I added it to nixpkgs: cfdd91c6fbdda90cd3725cb3aa2ffb5e8d8cbfd6

IvarWithoutBones commented:
Whoops, forgot to add a dash on one of the CMake flags. Also changed the order in `pkgs/top-level/all-packages.nix` to make alphabetical sense.

IvarWithoutBones commented:
Anything blocking this? 

As one of their core maintainers said [here](https://github.com/yuzu-emu/yuzu/issues/3066#issuecomment-568781233), their fork is considerably different. We won't be able to use Unicorn from nixpkgs.

jtojnar commented:
Not a strict blocker but ideally, we would figure out how does their unicorn fork differ and try to work with upstreams to accept the changes.

Also another contributor says it works with unicorn master.

IvarWithoutBones commented:
Please ignore the PR above, was a mistake.

unicorn master does not seem to work either though it gets a bit further:
```
[ 57%] Building CXX object src/core/CMakeFiles/core.dir/constants.cpp.o
/build/source/src/core/arm/unicorn/arm_unicorn.cpp: In static member function 'static void Core::ARM_Unicorn::InterruptHook(uc_engine*, u32, void*)':
/build/source/src/core/arm/unicorn/arm_unicorn.cpp:271:29: error: 'UC_ARM64_REG_ESR' was not declared in this scope; did you mean 'UC_ARM64_REG_LR'?
  271 |     CHECKED(uc_reg_read(uc, UC_ARM64_REG_ESR, &esr));
      |                             ^~~~~~~~~~~~~~~~
/build/source/src/core/arm/unicorn/arm_unicorn.cpp:36:27: note: in definition of macro 'CHECKED'
   36 |         if (auto _cerr = (expr)) {                                                                 \
      |                           ^~~~
make[2]: *** [src/core/CMakeFiles/core.dir/build.make:109: src/core/CMakeFiles/core.dir/arm/unicorn/arm_unicorn.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:1221: src/core/CMakeFiles/core.dir/all] Error 2
make: *** [Makefile:161: all] Error 2
builder for '/nix/store/27ixjh1pimisfsbwwm4k75a89f0v7j5n-yuzu-unstable-2020-04-25.drv' failed with exit code 2
error: build of '/nix/store/27ixjh1pimisfsbwwm4k75a89f0v7j5n-yuzu-unstable-2020-04-25.drv' failed
```

Missing function is added [here](https://github.com/yuzu-emu/unicorn/commit/73f45735354396766a4bfb26d0b96b06e5cf31b2) i think, i guess we should upstream that, or use it as a patch? Not sure if we can patch external packages from this one.

jtojnar commented:
Upstreaming is always good.

You can do something like

```nix
(unicorn.overrideAttrs (attrs: {
  patches = (attrs.patches or []) ++ [
    (fetchpatch {
      url = "https://github.com/yuzu-emu/unicorn/commit/73f45735354396766a4bfb26d0b96b06e5cf31b2.patch";
      sha256 = "0000000000000000000000000000000000000000000000000000000000000000";
    })
  ];
}))
```

IvarWithoutBones commented:
Should i reopen the PR linked above, or just patch everything the release doesn't have in? If we'd use master just this 1 patch does the trick, but with the stable release we need a fair few more.

jtojnar commented:
Would overriding the src here as well not be enough?

IvarWithoutBones commented:
If we fetch https://github.com/yuzu-emu/unicorn/tree/73f45735354396766a4bfb26d0b96b06e5cf31b2, yes. Somehow hadn't thought of that.

IvarWithoutBones commented:
Done, yuzu now correctly uses unicorn from nixpkgs :)

JoshuaFern commented:
A couple days ago I updated[ my personal version of Yuzu](https://github.com/JoshuaFern/nur-package-lab/blob/master/pkgs/misc/emulators/yuzu/default.nix) on the NUR, this was the most recent commit I could get building. Perhaps we should update the package to at least this commit?

Also, Yuzu has added Multicore emulation to their Early Access version, it allows most games to run at full speed by utilizing your processor more efficiently. I'm not sure when this feature will trickle down to mainline but it's a huge advancement that makes the emulator much more usable. We should again update as soon as it's available.

IvarWithoutBones commented:
Sounds good, I'll update to that rev. I've also gotten [some errors](https://gist.github.com/IvarWithoutBones/ec576fe8a863949c4f023a635e7ecfb2) trying to compile the most recent commit, regarding QT's webkit and FMT which I'm not sure how to fix. The problems seem to have arrived with their switch to Conan, which a lot of people seem to be having issues with which I'm sure they'll fix in the coming weeks. However for now we can just use the rev from your NUR package in my opinion, as we indeed want to update in a few weeks when multicore gets added to mainline anyway.

IvarWithoutBones commented:
Build errors seem to have been introduced with [this  PR](https://github.com/yuzu-emu/yuzu/pull/3735) as suspected. FMT issue seems to match [this](https://github.com/NixOS/nixpkgs/issues/86622) (and possibly [this](https://github.com/NixOS/nixpkgs/issues/87473)? Not sure), as for the QTWebKit one im not sure. I've also realized mainline does have releases, and we should probably use those instead of git revs. 249 is the last one i managed to get building, so I'll "update" to that.

maddovr commented:
The multicore version is still only EA only to iron out bugs, moreover, their vulkan backend is in a problematic state (massive memory leaks). This emulator is advancing very rapidly, honestly you'd realistically want to update it every week or so (Each month, we basically get new games being playable from start to finish on the emu).

maddovr commented:
My €0.02 is that it should be doable now since the stable branch is in quite good state atm (save vulkan, but I don't see it getting fixed before the end of the year honestly), the real issue is that you truly want to update yuzu **at least** bi-weekly, if not weekly, since new features and important bugfixes get merged basically daily from the early access builds back to the main branch, and stuff like compatibility grows by *everyday*. The trend is stable for now, and I'd expect a slow-down some time next year. Merging it once and forgetting to update it is basically the same as not merging at all.

Atemu commented:
As long as it's just version bumps, quick updates wouldn't be an issue. Besides, once it's in nixpkgs, you can easily override the src to get newer versions should you need it faster than nixpkgs can do it for you.

cleverca22 commented:
https://gist.github.com/e0dc5a0b9f0ede2366c393f36de914d0
recently wrote this, without knowing about this PR

it supports building latest master

IvarWithoutBones commented:
I've decided to pick this up again, and got the latest mainline to build :)
Everything seems to be working fine, based on a bit of testing.

This depends on https://github.com/NixOS/nixpkgs/pull/97266, which for now I've added using overrides.

aanderse commented:
I would love to see this get merged after #97266 is. I just did some testing on it and the package works great :+1:

IvarWithoutBones commented:
Now that #97266 is merged, I've updated the package to the latest version & removed the override for nlohmann_json.

Since my last comment, I've also enabled QT translations, as I noticed this was also enabled on their CI.

Should finally be ready to merge now :)

aanderse commented:
@IvarWithoutBones build seems to fail here:
```
[ 11%] Built target SoundTouch
[ 11%] Building C object externals/opus/CMakeFiles/opus.dir/opus/silk/HP_variable_cutoff.c.o
[ 11%] Building C object externals/mbedtls/library/CMakeFiles/mbedcrypto.dir/pkparse.c.o
In file included from /build/source/src/./common/logging/backend.h:10,
                 from /build/source/src/common/logging/filter.cpp:6:
/build/source/src/./common/file_util.h:60:38: error: wrong number of arguments specified for 'nodiscard' attribute
   60 |     const std::filesystem::path& path);
      |
```

aanderse commented:
I decided to confirm `gcc10` would fix this and it does. You can see the changes I applied against your branch [here](https://github.com/NixOS/nixpkgs/commit/debbc43aa24e45755f975d998f29cc8f57a6f3f5).

IvarWithoutBones commented:
Apologies for the late reply, got busy over the weekend.

With your diff applied, we're not using QT's `mkDerivation` anymore, but instead the `gcc10Stdenv` one. The big difference here being it not adding `wrapQtAppsHook` to the `nativeBuildInputs` anymore, however we can do this manually as well.

I think the solution I've just pushed is a bit clearer, as we just drop QT's `mkDerivation` all together, instead of overwriting it.

Everything is now running again, and I've updated to the latest version again as well :)

aanderse commented:
> I think the solution I've just pushed is a bit clearer, as we just drop QT's `mkDerivation` all together, instead of overwriting it.

Yes, definitely. I just wanted to confirm `gcc10` was the solution. Great stuff!

aanderse commented:
Anyone wanna merge this?

IvarWithoutBones commented:
/marvin opt-in
/status needs_reviewer

aanderse commented:
@jtojnar anything else before a merge?

kisik21 commented:
Tried to build it, turns out Nix thinks its unfree. Why is this package marked as unfree?

kisik21 commented:
I don't have a Switch so I can't verify if the software functions, but it launches.

/status needs_merger

aanderse reviewed (approved):
Merging based on:
- code LGTM :+1:
- approval by @jtojnar 
- **extensive** testing by my kids :laughing:

aanderse commented:
Thanks @IvarWithoutBones and all :tada:

IvarWithoutBones commented:
I'm_ really happy we finally managed to get this merged.
Thanks everyone for the extensive reviewing and helping out on this PR, I've learned a lot.


> Tried to build it, turns out Nix thinks its unfree. Why is this package marked as unfree?

@kisik21 Icons are licensed under `cc-by-nd-30`, which is unfree.
